File: c/1.1.5/1.1.5_4-Digit.c
/**********************************************************************
* Filename    : 1.1.5_4-Digit.c
* Description : Control 4 Digit 7-Segment Display
* Author      : Robot
* E-mail      : support@sunfounder.com
* website     : www.sunfounder.com
* Update      : Cavon    2016/07/01
**********************************************************************/
#include <wiringPi.h>
#include <stdio.h>
#include <wiringShift.h>
#include <signal.h>
#include <unistd.h>
#define     SDI     5   
#define     RCLK    4  
#define     SRCLK    1   
const int placePin[]={0,2,3,12};        // Define 4 digit's common pin

// A segment code array from 0 to 9 in Hexadecimal (Common anode) 
unsigned char number[]={0xc0,0xf9,0xa4,0xb0,0x99,0x92,0x82,0xf8,0x80,0x90};

int counter = 0;    //variable counter,the number will be displayed 

//select the place of the value, this is only one place that will be enable each time.
//the parameter digit is optional for 0001,0010,0100,1000
void selectPlace(int digit){    
    digitalWrite(placePin[0],!(digit&0x08));
    digitalWrite(placePin[1],!(digit&0x04));
    digitalWrite(placePin[2],!(digit&0x02));
    digitalWrite(placePin[3],!(digit&0x01));
}

void hc595_shift(int8_t data){      //To assign 8 bit value to 74HC595’s shift register

    int i;  
    for(i = 0; i < 8; i++){
        digitalWrite(SDI, 0x80 & (data << i));
        digitalWrite(SRCLK,1);
        delayMicroseconds(1);
        digitalWrite(SRCLK,0);
	}
        digitalWrite(RCLK,1);
        delayMicroseconds(1);
        digitalWrite(RCLK,0);        
}

void display(int counter){  //display the number
	hc595_shift(0xff);	
    selectPlace(0x01);      
    hc595_shift(number[counter%10]); //Display the number on the single digit of the value  
    delay(1); 
    
    hc595_shift(0xff);    
    selectPlace(0x02); 
    hc595_shift(number[counter%100/10]);
    delay(1);
    
    hc595_shift(0xff);    
    selectPlace(0x04); 
    hc595_shift(number[counter%1000/100]);
    delay(1);
    
    hc595_shift(0xff);    
    selectPlace(0x08);
    hc595_shift(number[counter%10000/1000]);
    delay(1);
}
void timer(int  sig){       //Timer function
    if(sig == SIGALRM){   //If the signal is SIGALRM, the value of counter plus 1, and update the number displayed by 7-segment display
        counter ++;         
        alarm(1);           //set the next timer time
        printf("counter : %d \n",counter);
    }
}
int main(void)
{
    int i;
    printf("\n");
	printf("========================================\n");
	printf("|        4 Digit 7-Segment Display     |\n");
	printf("|    ------------------------------    |\n");
	printf("|     4 Digit display common pin       |\n");
	printf("|         connect to #17 #27 #22 MOSI  |\n");
	printf("|     74HC595 CH_CP ST_CP DS pin       |\n");
	printf("|         connect to #18 #23 #24       |\n");
	printf("|                                      |\n");
	printf("|                            SunFounder|\n");
	printf("========================================\n");
	printf("\n");

    if(wiringPiSetup() == -1){ //when initialize wiring failed,print messageto screen
        printf("setup wiringPi failed !");
        return 1; 
    }
    pinMode(SDI,OUTPUT);        //set the pin connected to74HC595 for output mode
    pinMode(RCLK,OUTPUT);
    pinMode(SRCLK,OUTPUT);
    //set the pin connected to 7-segment display common end to output mode
    for(i=0;i<4;i++){       
        pinMode(placePin[i],OUTPUT);
        digitalWrite(placePin[i],HIGH);
    }
    signal(SIGALRM,timer);  //configure the timer
    alarm(1);               //set the time of timer to 1s
    while(1){
        display(counter);   //display the number counter
    }
    return 0;
}

File: python/1.1.5_4-Digit.py
#!/usr/bin/env python3

import RPi.GPIO as GPIO
import time
import threading

#define the pins connect to 74HC595
SDI   = 18      #serial data input(DS)
RCLK  = 16      #memory clock input(STCP)
SRCLK = 12       #shift register clock input(SHCP)
number = (0xc0,0xf9,0xa4,0xb0,0x99,0x92,0x82,0xf8,0x80,0x90)
# A segment code array from 0 to 9 in Hexadecimal (Common anode) 

placePin = (11,13,15,19)    # Define 4 digit's common pin
counter = 0         # Variable counter, the number will be displayed
t = 0               # define the Timer object

def setup():
    GPIO.setmode(GPIO.BOARD)    # Number GPIOs by its physical location
    GPIO.setup(SDI, GPIO.OUT)       # Set pin mode to output
    GPIO.setup(RCLK, GPIO.OUT)
    GPIO.setup(SRCLK, GPIO.OUT)
    for pin in placePin:
        GPIO.setup(pin,GPIO.OUT)
    
            
def hc595_shift(dat):      #shift the data to 74HC595
	for bit in range(0, 8):	
		GPIO.output(SDI, 0x80 & (dat << bit))
		GPIO.output(SRCLK, GPIO.HIGH)
		GPIO.output(SRCLK, GPIO.LOW)
	GPIO.output(RCLK, GPIO.HIGH)
	GPIO.output(RCLK, GPIO.LOW)

# select the place of the value, this is only one place that will be enable each time.
# the parameter digit is optional for 0001,0010,0100,1000    
def selectPlace(digit): 
    GPIO.output(placePin[0],GPIO.LOW if (digit&0x08) else GPIO.HIGH)
    GPIO.output(placePin[1],GPIO.LOW if (digit&0x04) else GPIO.HIGH)
    GPIO.output(placePin[2],GPIO.LOW if (digit&0x02) else GPIO.HIGH)
    GPIO.output(placePin[3],GPIO.LOW if (digit&0x01) else GPIO.HIGH)

def display(counter):   #display the number
    hc595_shift(0xff)   #clean up the display
    selectPlace(0x01)   #Select place
    hc595_shift(number[counter%10]) #Display the number on the single digit of the value  
    time.sleep(0.003)   
    hc595_shift(0xff)
    selectPlace(0x02) 
    hc595_shift(number[counter%100//10])
    time.sleep(0.003)
    hc595_shift(0xff)
    selectPlace(0x04)  
    hc595_shift(number[counter%1000//100])
    time.sleep(0.003)
    hc595_shift(0xff)
    selectPlace(0x08) 
    hc595_shift(number[counter%10000//1000])
    time.sleep(0.003)

def timer():        #timer function
    global counter
    global t
    t = threading.Timer(1.0,timer)      #set the time again
    t.start()                           #Start timing
    counter+=1                          
    print ("counter : %d"%counter)
    
def loop():
    global t
    global counter
    t = threading.Timer(1.0,timer)      #set the timer
    t.start()                           # Start timing
    while True:
        display(counter)                # display the number counter
        
def destroy():   # When "Ctrl+C" is pressed, the function is executed. 
    global t
    GPIO.cleanup()      
    t.cancel()      #cancel the timer

if __name__ == '__main__': # Program starting from here 
    setup() 
    try:
        loop()  
    except KeyboardInterrupt:  
        destroy()  
 
